### ISSUE TITLE
A11y: Setting to require alt text on all images in a post before publishing

### ISSUE THREAD

**Issue by cagrimmett** [WordPress/gutenberg#59754]
## What problem does this address?

Alt text often easily forgotten, even by a11y-minded users. 

## What is your proposed solution?

Bluesky has a setting that, when enabled, prevents a post from being published unless each image in the post has alt text added. This is something that should be adopted in Gutenberg. The setting can live and be toggled on/off in Options > Preferences > Accessibility.

![CleanShot 2024-03-11 at 09 56 00@2x](https://github.com/WordPress/gutenberg/assets/8738771/993998ce-d790-4ca7-bcd3-b6c0393f017f)

If a user tries to publish a post and has images without alt text added, the Publish button can be disabled and a warning can be shown at the top of the editor settings panel.

**Comment by cagrimmett** [comment]
The relevant discussion for the feature over in the Bluesky app repo: https://github.com/bluesky-social/social-app/issues/907

**Comment by joedolson** [comment]
Overall, I definitely support this. There are some caveats, however.

First, there are legitimate cases where an image may not need alt text; so we should allow a person to explicitly mark an image as decorative. One example is a case where a link is wrapping both an image and text. 

Second, could we to add some basic qualifiers to assess whether or not the requirement is fulfilled. It's probably beyond what we can do at this time to have any real quality assessment of alt text, but there are some obvious patterns that we might want to check for, e.g.

- Alt text is same as file name
- Alt text is a file name like string, e.g. "IMG_nnnn" or "DSC_nnnn"
- Alt text is a non-text string, e.g. `""`, `--`, or `&nbsp;`
- Alt text is very long
- Alt text is very short
- Alt text contains words like 'image', 'picture', or 'logo'. 

I'd like to see the pre-publish screen be able to flag suspicious alt text and raise that as warnings. They should be something that a user can bypass, as even these suspicious cases *could* be valid, but a double check is valuable. 

I've got a variety of simple checks that I'm doing in my plugin WP Accessibility, for reference: https://github.com/joedolson/wp-accessibility/blob/master/src/wp-accessibility-alt.php

**Comment by skorasaurus** [comment]
Heavily overlaps with https://github.com/WordPress/gutenberg/issues/16909 which proposes to prompt for alt-text immediately after inserting an image

**Comment by cbirdsong** [comment]
There is prior work around this sort of feature in the PublishPress Checklists plugin: 
- https://publishpress.com/checklists/
- https://wordpress.org/plugins/publishpress-checklists/
- https://github.com/publishpress/PublishPress-Checklists

- - -

> First, there are legitimate cases where an image may not need alt text; so we should allow a person to explicitly mark an image as decorative. One example is a case where a link is wrapping both an image and text.

This, very much. An explicit “this image is only used for presentation” toggle would be fantastic since it would allow the markup to more accurately reflect the intent of the author. As it is right now, `alt=“”` is more or less part of the default output for images - I’m not sure it’s even _possible_ to exclude it. This means tooling like PublishPress Checklists and [Editoria11y](https://wordpress.org/plugins/editoria11y-accessibility-checker/) can’t show definitive error messages[^block] for images with `alt=“”` and instead have to display mushy qualified warnings explaining the situation and leaving it up to the user.

With this option, image markup generated by the editor could exclude the `alt` attribute entirely unless alt text has been written or the image has been marked as decorative, making it easier for tooling inside and outside of WordPress to flag incorrect images more reliably.

[^block]: or fully block publishing of the post, which I think should definitely be an option site owners can enable even if it’s not the default

**Comment by joedolson** [comment]
We don't want to exclude `alt=""` - having an alt attribute present is very important. Having it not populated is *completely different* from having it absent.

If the alt attribute is absent, screen readers have to guess whether or not the image is intended as content, and will attempt to communicate some information about the image. Different screen readers will return different results, depending on many factors. 

With `alt=""` the image will be silenced, and this will be universal and expected.

Marking an image as decorative is an internal declaration that confirms user intent and indicates that the image with `alt=""` is acceptable. We may want to consider adding a data attribute for our own use, so that we can recognize those decisions.

**Comment by cbirdsong** [comment]
Ah, I didn't know having it missing had so many oddities. Adding `role="presentation"` might do the trick, though that may also have issues if the image is inside a link or button?

**Comment by joedolson** [comment]
Would probably need to use `role="none presentation"`, since `none` is the preferred value as of ARIA 1.1, but may not have complete support.

Overall, `alt=""` has complete support; alternatives would require some testing to verify. 

If an image is the sole content of a link or a button, then it should have alt text, so it would be good to consider that case.
